pkg/retriever/retriever.go

```diff
@@ -90,26 +90,29 @@ func NewRetriever(
 	config RetrieverConfig,
 	client RetrievalClient,
 	candidateFinder CandidateFinder,
+	bitswapRetriever types.CandidateRetriever,
 ) (*Retriever, error) {
 	retriever := &Retriever{
 		config:       config,
 		eventManager: events.NewEventManager(ctx),
 		spTracker:    newSpTracker(nil),
 	}
+	candidateRetrievers := []types.CandidateRetriever{
+		&GraphSyncRetriever{
+			GetStorageProviderTimeout: retriever.getStorageProviderTimeout,
+			IsAcceptableQueryResponse: retriever.isAcceptableQueryResponse,
+			Client:                    client,
+		},
+	}
+	if bitswapRetriever != nil {
+		candidateRetrievers = append(candidateRetrievers, candidateRetrievers...)
+	}
 	retriever.executor = combinators.RetrieverWithCandidateFinder{
 		CandidateFinder: NewAssignableCandidateFinder(candidateFinder, retriever.isAcceptableStorageProvider),
 		CandidateRetriever: combinators.SplitRetriever{
-			CandidateSplitter: NewProtocolSplitter([]multicodec.Code{multicodec.TransportGraphsyncFilecoinv1, multicodec.TransportBitswap}),
-			CandidateRetrievers: []types.CandidateRetriever{
-				&GraphSyncRetriever{
-					GetStorageProviderTimeout:   retriever.getStorageProviderTimeout,
-					IsAcceptableStorageProvider: retriever.isAcceptableStorageProvider,
-					IsAcceptableQueryResponse:   retriever.isAcceptableQueryResponse,
-					Client:                      client,
-				},
-				NewBitswapRetriever(),
-			},
-			CoordinationKind: types.RaceCoordination,
+			CandidateSplitter:   NewProtocolSplitter([]multicodec.Code{multicodec.TransportGraphsyncFilecoinv1, multicodec.TransportBitswap}),
+			CandidateRetrievers: candidateRetrievers,
+			CoordinationKind:    types.RaceCoordination,
```

Comment:
This is going to get weird if/when we have both strategies using the same linksystem.
The CAR(s) shouldn't double-write if both bitswap and graphsync both fetch the same block and try and put it, but we ought to test that this works cleanly and we don't end up with weird CARs.

Reply:
yea I know we'll fix in the streaming PR